### PR TITLE
Ability for log to langchain as code

### PR DESCRIPTION
**Pull request by sunishsheth2009**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/sunishsheth2009/mlflow/pull/11370?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11370/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11370
```

</p>
</details>

### What changes are proposed in this pull request?
Ability for log to langchain as code

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/langchain/__init__.py [hunk 1]
```diff
@@ -230,7 +243,12 @@ def load_retriever(persist_directory):
 
     path = os.path.abspath(path)
     _validate_and_prepare_target_save_path(path)
-    code_dir_subpath = _validate_and_copy_code_paths(code_paths, path)
+    formatted_code_path = (
+        [lc_model]
+        if isinstance(lc_model, str) and code_paths is None
+        else (code_paths or []) + ([lc_model] if isinstance(lc_model, str) else [])
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:249]:
```suggestion
    code_paths = code_paths or []
    if isinstance(lc_model, str):
        # The LangChain model is defined as Python code located in the file at the path
        # specified by `lc_model`. Verify that the path exists and, if so, copy it to the
        # model directory along with any other specified code modules
        
        if os.path.exists(lc_model):
            code_paths.append(lc_model)
        else:
            raise MlflowException(...)
```

Comment (dbczumar) [reply mlflow/langchain/__init__.py:249]: Can we also add a test verifying that an exception is thrown if the specified path doesn't exist?

File: mlflow/langchain/__init__.py [hunk 2]
```diff
@@ -276,7 +294,17 @@ def load_retriever(persist_directory):
...
-    model_data_kwargs = _save_model(lc_model, path, loader_fn, persist_dir)
+    if not isinstance(lc_model, str):
+        model_data_kwargs = _save_model(lc_model, path, loader_fn, persist_dir)
+        flavor_conf = {
+            _MODEL_TYPE_KEY: lc_model.__class__.__name__,
+            **model_data_kwargs,
+        }
+    else:
+        # if the model is a string, we only expect one code_path for now, which is config yaml file
+        file = os.path.basename(os.path.abspath(code_paths[0]))
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:305]: But we just appended to `code_paths` above, so the length will be `2`, not `1`. This is somewhat hard to follow.

Also, doesn't this indicate that lines 248 / 249 support an incorrect case - `code_paths` should never be `None` or empty if `isinstance(lc_model, str)` is True. `config.yml` always needs to be specified, right? Can we add a check around 248 / 249 and throw if this isn't true?

File: mlflow/langchain/__init__.py [hunk 3]
```diff
@@ -276,7 +294,17 @@ def load_retriever(persist_directory):
...
+        flavor_conf = {_DATABRICKS_RAG_CONFIG: file}
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:306]: Can we add a comment explaining what's going on here? I.e. why are we setting this RAG conf and what's it for?

File: mlflow/langchain/__init__.py [hunk 4]
```diff
@@ -286,13 +314,15 @@ def load_retriever(persist_directory):
...
-    }
 
     if Version(langchain.__version__) >= Version("0.0.311"):
-        if databricks_dependency := _detect_databricks_dependencies(lc_model):
+        if isinstance(lc_model, str):
+            spec = importlib.util.spec_from_file_location("temp_module", lc_model)
+            temp_module = importlib.util.module_from_spec(spec)
+            spec.loader.exec_module(temp_module)
+
+        checker_model = lc_model if lc_model is not None else __chain__
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:324]: Can we move this into a separate method called `_load_model_from_code`?

File: mlflow/langchain/utils.py [hunk 5]
```diff
@@ -263,6 +264,13 @@ def _validate_and_wrap_lc_model(lc_model, loader_fn):
     import langchain.llms.openai
     import langchain.schema
 
+    if isinstance(lc_model, str):
+        if os.path.basename(os.path.abspath(lc_model)) != "chain.py":
+            raise mlflow.MlflowException.invalid_parameter_value(
+                f"The path to the chain.py file must be provided, but provided {lc_model}."
```
Comment (dbczumar) [inline mlflow/langchain/utils.py:270]: Can we be a bit more explicit and say something like "If `lc_model` is a string, it must be a path to a file called `chain.py` on the local filesystem?"

File: mlflow/langchain/__init__.py [hunk 6]
```diff
@@ -84,6 +86,17 @@
 
 FLAVOR_NAME = "langchain"
 _MODEL_TYPE_KEY = "model_type"
+_DATABRICKS_RAG_CONFIG = "databricks_rag_config"
+__chain__ = None
+__config_path__ = None
+
+
+def _set_chain(chain):
+    globals()["__chain__"] = chain
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:95]: This isn't used. Should it be here, or should it be defined in another module like `databricks.rag`?

File: mlflow/langchain/__init__.py [hunk 7]
```diff
@@ -84,6 +86,17 @@
...
+
+
+def _set_config_path(path):
+    globals()["__config_path__"] = path
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:99]: Should this be defined in `databricks.rag` instead?

**Review by dbczumar — COMMENTED** [review]
Thanks @sunishsheth2009  - left some comments! Can we also add test coverage verifying that these code-based models can be logged (including correct inferred Databricks dependency info like vector stores), loaded, and served?

File: mlflow/langchain/__init__.py (hunk 6 above)
Comment (sunishsheth2009) [reply mlflow/langchain/__init__.py:97]: We need that users in MLflow folks can do that and also from databricks.rag we import mlflow.langchain and set that as well. So its reused for that.

File: mlflow/langchain/__init__.py (hunk 7 above)
Comment (sunishsheth2009) [reply mlflow/langchain/__init__.py]: We need that users in MLflow folks and it is very weird that mlflow calls databricks.rag and create a cyclic dependency. So databricks.rag imports __config_path__ to get that information. 

Also made it more generic. Let me know if that makes sense. :)

File: mlflow/langchain/__init__.py (hunk 6 above)
Comment (dbczumar) [reply mlflow/langchain/__init__.py:95]: This is only for Databricks RAG, not MLflow users. Do we need to put this in mlflow.langchain for the Databricks RAG use case?

Comment (sunishsheth2009) [reply mlflow/langchain/__init__.py:95]: I think it would be easier for databrick.rag to depend on mlflow.langchain rather than creating a circular dependency. What do you think?

Comment (dbczumar) [reply mlflow/langchain/__init__.py:95]: I'd rather isolate this funny business to `databricks.rag` for now. It's a bit odd to bleed it into the langchain package.

File: mlflow/langchain/__init__.py [hunk 8]
```diff
@@ -84,6 +86,17 @@
 
 FLAVOR_NAME = "langchain"
 _MODEL_TYPE_KEY = "model_type"
+_CODE_CONFIG = "_code_config"
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:89]: Can we add an inline comment explaining what this is used for?

File: mlflow/langchain/__init__.py [hunk 9]
```diff
@@ -230,7 +243,19 @@ def load_retriever(persist_directory):
...
+    formatted_code_path = code_paths[:] if code_paths else []
+    if isinstance(lc_model, str):
+        # The LangChain model is defined as Python code located in the file at the path
+        # specified by `lc_model`. Verify that the path exists and, if so, copy it to the
+        # model directory along with any other specified code modules
+
+        if os.path.exists(lc_model):
+            formatted_code_path.append(lc_model)
+        else:
+            raise mlflow.MlflowException(
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:255]: Can we use `INVALID_PARAMETER_VALUE` for the exception code for now?

File: mlflow/langchain/__init__.py [hunk 10]
```diff
@@ -230,7 +243,19 @@ def load_retriever(persist_directory):
...
+                f"If the {lc_model} is a string, it must be a valid file path."
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:256]: Can we be more specific about what the file should be? I.e. a Python file that defines a LangChain chain.

File: mlflow/langchain/__init__.py [hunk 11]
```diff
@@ -286,13 +322,15 @@ def load_retriever(persist_directory):
...
-        _MODEL_TYPE_KEY: lc_model.__class__.__name__,
-        **model_data_kwargs,
-    }
 
     if Version(langchain.__version__) >= Version("0.0.311"):
-        if databricks_dependency := _detect_databricks_dependencies(lc_model):
+        checker_model = lc_model
+        if isinstance(lc_model, str):
+            _add_code_to_system_path(lc_model)
+            _load_code_model("", flavor_conf, True)
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:330]: Passing an empty string here is a little odd, and it's also a bit strange to pass the flavor conf when we already have the code path that the flavor conf refers to. Can we refactor `_load_code_model` so that it takes a code path directly and then do the path creation here and in `load_model()`?

Comment (dbczumar) [reply mlflow/langchain/__init__.py:330]: While we're at it, can we remove the `isLocal` logic too? Seems unnecessary if we construct the path before passing it to `_load_code_model`.

File: mlflow/langchain/__init__.py [hunk 12]
```diff
@@ -286,13 +322,15 @@ def load_retriever(persist_directory):
...
-    flavor_conf = {
-        _MODEL_TYPE_KEY: lc_model.__class__.__name__,
-        **model_data_kwargs,
-    }
 
     if Version(langchain.__version__) >= Version("0.0.311"):
-        if databricks_dependency := _detect_databricks_dependencies(lc_model):
+        checker_model = lc_model
+        if isinstance(lc_model, str):
+            _add_code_to_system_path(lc_model)
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:329]: Shouldn't this be called within `_load_code_model`?

File: mlflow/langchain/__init__.py [hunk 13]
```diff
@@ -726,6 +767,24 @@ def load_model(model_uri, dst_path=None):
...
+        path: os.path.join(
+            local_model_path,
+            flavor_code_config,
+            path if isLocal else os.path.basename(os.path.abspath(path)),
+        )
+        for path in databricks_rag_config
+    }
+    _set_config_path(code_paths)
+
+    import chain  # noqa: F401
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:783]: I would expect to add the code path to the system path before performing this import. I think that should be done within this method. See also https://github.com/mlflow/mlflow/pull/11370/files#r1520525504.

File: mlflow/langchain/utils.py [hunk 14]
```diff
@@ -263,6 +264,13 @@ def _validate_and_wrap_lc_model(lc_model, loader_fn):
     import langchain.llms.openai
     import langchain.schema
 
+    if isinstance(lc_model, str):
+        if os.path.basename(os.path.abspath(lc_model)) != "chain.py":
+            raise mlflow.MlflowException.invalid_parameter_value(
+                f"If {lc_model} is a string, it must be called chain.py on the local filesystem"
```
Comment (dbczumar) [inline mlflow/langchain/utils.py:270]:
```suggestion
                f"If {lc_model} is a string, it must be the path to a file named `chain.py` on the local filesystem."
```

File: mlflow/langchain/__init__.py (hunk 12 above)
Comment (sunishsheth2009) [reply mlflow/langchain/__init__.py:329]: Here we are just adding lc_model to sys.path which is needed just for this usecase. 

In loading the model back, we are adding the entire local_path to sys path: https://github.com/mlflow/mlflow/pull/11370/files/06759dae66a46a4fb753f6e72612a37753585c83#diff-44b09327d1292c38e7ce251f66635a35c22720697ae1f39db6d4873d98109548R736

File: mlflow/langchain/__init__.py (hunk 13 above)
Comment (sunishsheth2009) [reply mlflow/langchain/__init__.py:783]: In loading the model back, we are adding the entire local_path to sys path: https://github.com/mlflow/mlflow/pull/11370/files/06759dae66a46a4fb753f6e72612a37753585c83#diff-44b09327d1292c38e7ce251f66635a35c22720697ae1f39db6d4873d98109548R736

File: mlflow/langchain/__init__.py [hunk 15]
```diff
@@ -85,6 +87,15 @@
...
 _MODEL_TYPE_KEY = "model_type"
 
+# The following constants is used in flavor configuration to specify the code paths
+# that should be set using _set_config_path when the model is loaded by code.
+_CODE_CONFIG = "_code_config"
+__code_paths__ = None
+
+
+def _set_config_path(path):
+    globals()["__code_paths__"] = path
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:97]: Can we move this to `_rag_utils` as well?

File: mlflow/langchain/_rag_utils.py [hunk 16]
```diff
@@ -0,0 +1,5 @@
+__chain__ = None
+
+
+def _set_chain(chain):
```
Comment (dbczumar) [inline mlflow/langchain/_rag_utils.py:4]: Can we add a docstring to this method explaining why it's here / what it does / what the scope is? Can we do the same for `_set_config_path` once we move it here?

File: mlflow/langchain/__init__.py (hunk 12 above)
Comment (dbczumar) [reply mlflow/langchain/__init__.py:330]: Hmmm...I think it's a bit odd for `_load_code_model` not to handle the system pathing - can we leave a comment in the code explaining why we're doing this?

File: mlflow/langchain/__init__.py (hunk 11 above)
Comment (dbczumar) [reply mlflow/langchain/__init__.py]: @sunishsheth2009 Looks like we didn't fix this

File: mlflow/langchain/__init__.py [hunk 17]
```diff
@@ -722,10 +782,20 @@ def load_model(model_uri, dst_path=None):
...
     """
-    local_model_path = _download_artifact_from_uri(artifact_uri=model_uri, output_path=dst_path)
+    local_model_path = _download_artifact_from_uri(
+        artifact_uri=model_uri, output_path=dst_path
+    )
     return _load_model_from_local_fs(local_model_path)
 
 
+def _load_code_model(code_paths):
+    _set_config_path(code_paths)
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:792]: We're calling the argument `code_paths`, but `set_config_path` assumes that this is a single path, not a list of paths. Can we call this `config_path`? Callers should also verify that `code_paths` contains a single entry before calling `_load_code_model`.

Comment (dbczumar) [reply mlflow/langchain/__init__.py:792]: Actually, why does this work with https://github.com/mlflow/mlflow/pull/11370/files#diff-44b09327d1292c38e7ce251f66635a35c22720697ae1f39db6d4873d98109548R331? Seems like we're erroneously passing in a set here, instead of a string.

File: mlflow/langchain/__init__.py (hunk 15 above)
Comment (dbczumar) [inline mlflow/langchain/__init__.py:97]: Shouldn't this be called `config_path`?

```suggestion
    globals()["__databricks_rag_config_path__"] = path
```

File: mlflow/langchain/_rag_utils.py [hunk 18]
```diff
@@ -0,0 +1,5 @@
...
+    globals()["__chain__"] = chain
```
Comment (dbczumar) [inline mlflow/langchain/_rag_utils.py:5]:
```suggestion
    globals()["__databricks_rag_chain__"] = chain
```

File: tests/langchain/chain.py [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+
+        @property
+        def _llm_type(self) -> str:
+            return "fake chat model"
+
+    return FakeChatModel(endpoint=endpoint)
+
+
+text_path = "tests/langchain/state_of_the_union.txt"
+loader = TextLoader(__code_paths__.get(text_path, text_path))
```
Comment (dbczumar) [inline tests/langchain/chain.py:55]: This isn't how RagConfig would interact with `__code_paths__`, right? Wouldn't it expect a single path referring to `config.yaml`?

Comment (dbczumar) [inline tests/langchain/chain.py:55]: @sunishsheth2009 The current test is passing `text_path` as a default value if `text_path` is missing from the dict. This means we aren't verifying that `__code_paths__` contains the correct content.

```suggestion
loader = TextLoader(__code_paths__.get(text_path))
```

Comment (dbczumar) [reply tests/langchain/chain.py:55]: Can we also add an `assert` on the contents of `__code_paths__`?

File: mlflow/langchain/__init__.py (hunk 11 above)
Comment (sunishsheth2009) [reply mlflow/langchain/__init__.py]: https://databricks.slack.com/archives/C06KX8NAV8D/p1710207768495239?thread_ts=1710196888.373229&cid=C06KX8NAV8D
Discussed offline that this is already resolved above ^^

File: mlflow/langchain/__init__.py [hunk 20]
```diff
@@ -286,13 +319,22 @@ def load_retriever(persist_directory):
...
 
     if Version(langchain.__version__) >= Version("0.0.311"):
-        if databricks_dependency := _detect_databricks_dependencies(lc_model):
+        checker_model = lc_model
+        if isinstance(lc_model, str):
+            # If the model is a string, we are adding the model code path to the system path
+            # so it can be loaded correctly.
+            _add_code_to_system_path(lc_model)
+            if len(code_paths) > 1:
+                raise mlflow.MlflowException(
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:330]: Nit: Can we use invalid_parameter_value error code?

File: mlflow/langchain/__init__.py [hunk 21]
```diff
@@ -286,13 +319,22 @@ def load_retriever(persist_directory):
...
+                    "When the model is a string, only one code path is allowed."
+                    "Current code paths: {code_paths}"
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:332]: Can we explain what that one code path is supposed to be?

File: mlflow/langchain/__init__.py [hunk 22]
```diff
@@ -85,6 +88,10 @@
 FLAVOR_NAME = "langchain"
 _MODEL_TYPE_KEY = "model_type"
 
+# The following constants is used in flavor configuration to specify the code paths
+# that should be set using _set_code_path when the model is loaded by code.
+_CODE_CONFIG = "_code_config"
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:93]: Can we move this to `_rag_utils`  update the comment to indicate that it's RAG-specific?

File: mlflow/langchain/__init__.py [hunk 23]
```diff
@@ -286,13 +319,22 @@ def load_retriever(persist_directory):
...
+        if isinstance(lc_model, str):
+            # If the model is a string, we are adding the model code path to the system path
+            # so it can be loaded correctly.
+            _add_code_to_system_path(lc_model)
+            if len(code_paths) > 1:
+                raise mlflow.MlflowException(
+                    "When the model is a string, only one code path is allowed."
+                    "Current code paths: {code_paths}"
+                )
+            _load_code_model(code_paths[0])
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:334]: Where do we verify that `code_paths` is not empty? We should do that before attempting to access the first element.

File: mlflow/langchain/_rag_utils.py [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+def _set_chain(chain):
+    """
+    The function is used to set the chain by the users of
+    the library at a global level. This global state then
+    can be referenced by the library to get the chain instance.
+    """
+    globals()["__databricks_rag_chain__"] = chain
+
+
+def _set_code_path(path):
```
Comment (dbczumar) [inline mlflow/langchain/_rag_utils.py:14]:
```suggestion
def _set_config_path(path):
```

File: mlflow/langchain/_rag_utils.py [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+    """
+    The function is used by the library to provide the local
+    path of the code path to the users so it can be referenced
+    while loading the chain back.
```
Comment (dbczumar) [inline mlflow/langchain/_rag_utils.py:18]: Can we be more specific? This is for RAG on Databricks.

File: mlflow/langchain/_rag_utils.py [hunk 26]
```diff
@@ -0,0 +1,20 @@
+__databricks_rag_chain__ = None
+__databricks_rag_config_path__ = None
+
+
+def _set_chain(chain):
+    """
+    The function is used to set the chain by the users of
+    the library at a global level. This global state then
+    can be referenced by the library to get the chain instance.
+    """
```
Comment (dbczumar) [inline mlflow/langchain/_rag_utils.py:10]: Can we be more specific? This is for the Databricks RAG library

File: mlflow/langchain/__init__.py [hunk 27]
```diff
@@ -276,7 +298,18 @@ def load_retriever(persist_directory):
...
+        model_data_kwargs = _save_model(lc_model, path, loader_fn, persist_dir)
+        flavor_conf = {
+            _MODEL_TYPE_KEY: lc_model.__class__.__name__,
+            **model_data_kwargs,
+        }
+    else:
+        # If the model is a string, we expect other files that would be used in the model.
+        # We set the other paths here so they can be set globally when the model is loaded.
+        # with the local path. So the consumer can use that path.
+        flavor_conf = {_CODE_CONFIG: code_paths[0]}
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:311]: Same comment as https://github.com/mlflow/mlflow/pull/11370/files#r1520807748 - can we verify that code_paths contains exactly one element somewhere?

File: tests/langchain/chain.py [hunk 28]
```diff
@@ -0,0 +1,80 @@
...
+
+        @property
+        def _llm_type(self) -> str:
+            return "fake chat model"
+
+    return FakeChatModel(endpoint=endpoint)
+
+
+path = "tests/langchain/state_of_the_union.txt"
+text_path = mlflow.langchain._rag_utils.__databricks_rag_config_path__ or path
```
Comment (dbczumar) [inline tests/langchain/chain.py:55]: This file is always going to exist on the local filesystem, so if `__databricks_rag_config_path__` is `None`, this test won't catch it. Can we:

1. Create a new file in a temporary directory and have this script reference it by environment variable, e.g.

```
text_path = os.environ[...] or mlflow.langchain._rag_utils.__databricks_rag_config_path__
```

3. After the model packaging step, delete the file before the model is loaded / served

Comment (sunishsheth2009) [reply tests/langchain/chain.py:55]: I completely removed setting the path `state_of_the_union.txt` to path since it should always use `mlflow.langchain._rag_utils.__databricks_rag_config_path__`. So now we can make sure it will always exists.

**Comment by sunishsheth2009** [comment]
@mlflow-automation autoformat

File: mlflow/langchain/__init__.py [hunk 29]
```diff
@@ -230,12 +233,34 @@ def load_retriever(persist_directory):
...
+        else:
+            raise mlflow.MlflowException.invalid_parameter_value(
+                f"If the {lc_model} is a string, it must be a valid python "
+                "file path containing the code for defining the chain instance."
+            )
+
+        if len(code_paths) != 1:
+            raise mlflow.MlflowException.invalid_parameter_value(
+                "When the model is a string, there should be a config path provided. "
+                "This config path is used to set config.yml file path "
```
Comment (dbczumar) [inline mlflow/langchain/__init__.py:253]: Nit: Can we mention that the path must be specified using `code_paths` and that only one path can be specified?

**Comment by dbczumar** [comment]
@mlflow-automation autoformat

**Review by dbczumar — APPROVED** [review]
LGTM with one nit - thanks @sunishsheth2009 !
